src/codegen/program.lisp

```diff
@@ -209,7 +208,19 @@ Example:
   (append
    ;; Predeclare symbol macros
    (loop :for (name . node) :in bindings
-         :collect `(global-lexical:define-global-lexical ,name ,(tc:lisp-type (node-type node) env)))
+         :append
+         `((global-lexical:define-global-lexical ,name ,(tc:lisp-type (node-type node) env))
+           ,@(when (and (node-abstraction-p node)
```

Comment:
Is there a reason to not put this logic below, just above the `compile-function` call?

Also, please add a comment above the declaration saying why we do an external declaim over an internal declare.

Reply:
This is to make sure that all `declaim`s appear before function definitions.
It is the point of this change to optimize tail-calls with SBCL.

For example, the following code won't be optimized and will raise call stack exhausted with a large number.

```common-lisp
(declaim isodd? ...)
(defun isodd? (n)
  ;; Calls iseven?
  )

(declaim iseven? ...)
(defun iseven? (n)
  ;; Calls isodd?
  )
```

However, it works with this.

```common-lisp
;; Definitions first
(declaim isodd? ...)
(declaim iseven? ...)

(defun isodd? (n)
  ;; Calls iseven?
  )
(defun iseven? (n)
  ;; Calls isodd?
  )
```

I'll add a comment.

Reply:
I added a comment.
I moved the code into an individual loop rather than sharing with symbol macros.